src/ucs/config/parser.c

```diff
@@ -1066,6 +1066,173 @@ void ucs_config_help_table(char *buf, size_t max, const void *arg)
+            if (!strcmp(cfg_key->name, key)) {
+                values[index] = value;
+                return UCS_OK;
+            }
+        }
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+int ucs_config_sscanf_key_value(const char *buf, void *dest, const void *arg)
+{
+    const ucs_config_key_value_param_t *param = arg;
+    size_t buf_len                            = strlen(buf) + 1;
+    const char *TOKEN_DELIM                   = ",";
+    const char *VALUE_DELIM                   = ":";
+    size_t key_count, index;
+    const char **values;
+    char *token, *tokens;
+    const char *key, *value;
+
+    /* values is a temporary array value storage for the keys.
+     * In values array we remember the last applicable setting for each key.
+     * We need it to perform actual setting exactly once per each key */
+    key_count = ucs_config_key_value_count(param->keys);
+    values = ucs_alloca(key_count * sizeof(*values));
+    memset(values, 0, key_count * sizeof(*values));
+
+    /* Writable copy of the input buffer on stack */
+    tokens = ucs_alloca(buf_len);
+    ucs_strncpy_safe(tokens, buf, buf_len);
+
+    tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
+    while (NULL != token) {
+        ucs_string_split(token, VALUE_DELIM, 2, &key, &value);
+
+        /* Remember last value in the values array */
+        if (UCS_OK != ucs_config_key_value_set(param->keys, values, key_count,
+                                               key, value)) {
+            ucs_error("Key '%s' is not supported", key);
+            return 0;
+        }
+        tokens = ucs_string_split(tokens, TOKEN_DELIM, 1, &token);
+    }
+
+    /* Now apply stored temporary values, exactly once per each key */
+    for (index = 0; index < key_count; ++index) {
+        if (NULL == values[index]) {
+            ucs_error("No value configured for key '%s'",
+                      param->keys[index].name);
+            return 0;
+        }
+        if (param->parser.read(values[index], dest + param->keys[index].offset,
+                               param->parser.arg) != 1) {
+            /* Error is logged by parser impl */
+            return 0;
+        }
+    }
+    return 1;
+}
+
+int ucs_config_sprintf_key_value(char *buf, size_t max, const void *src,
+                                 const void *arg)
+{
+    const ucs_config_key_value_param_t *param = arg;
+    char value_buf[256]                       = "";
+    const ucs_config_key_field_t *cfg_key;
+    ucs_string_buffer_t strb;
+
+    ucs_string_buffer_init_fixed(&strb, buf, max);
+
+    for (cfg_key = param->keys; NULL != cfg_key->name; ++cfg_key) {
+        if (!param->parser.write(value_buf, sizeof(value_buf),
+                                 src + cfg_key->offset, param->parser.arg)) {
+            /* Error is logged by parser impl */
+            return 0;
+        }
+        if (cfg_key != param->keys) {
+            ucs_string_buffer_appendf(&strb, ",");
+        }
```

Comment:
can add "," every time, and then use ucs_string_buffer_rtrim

Reply:
thanks, good idea indeed